src/libtorch.cc

```diff
@@ -512,6 +545,15 @@ ModelInstanceState::ModelInstanceState(
 
 ModelInstanceState::~ModelInstanceState()
 {
+  // If this is the last instance, remove the pointer from the model map
+  if (torch_model_.use_count() == 2) {
+    if (StateForModel()->UnloadModel(
+            std::make_pair(!device_.is_cpu(), device_.index())) != 1) {
+      LOG_MESSAGE(
+          TRITONSERVER_LOG_ERROR,
+          ("Model mapping not found for '" + Name() + "'").c_str());
+    }
+  }
   torch_model_.reset();
```

Comment:
Does this not reset the shared pointer for all model instances on the same device?

Reply:
No. It removes the current instance of shared pointer. When the last shared pointer is reset, the memory is deallocated.

I confirmed this via logging (using ::use_count to see the number of shared pointers) and C++ documentation: https://en.cppreference.com/w/cpp/memory/shared_ptr/reset
"If *this already owns an object and it is the last shared_ptr owning it, the object is destroyed through the owned deleter."